### PR TITLE
Bullet collision tuning

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

Various small fixes to improve runtime perf related to Bullet collision-detection. For details, see inline comments.

One change here is an overhaul of collision filtering. This is the choice of "group" and "mask" for collision objects added to the Bullet world. My new proposed scheme is:

- Static objects use group btBroadphaseProxy::StaticFilter and are set to collide with dynamic objects.
- Dynamic objects use group btBroadphaseProxy::DefaultFilter and are set to collide with static objects, kinematic objects, and other dynamic objects.
- Kinematic objects use group btBroadphaseProxy::KinematicFilter and are set to collide with dynamic objects.
- castRay uses group btBroadphaseProxy::DefaultFilter and is set to collide with everything (btBroadphaseProxy::AllFilter). This is the default Bullet behavior set in src/BulletCollision/CollisionDispatch/btCollisionWorld.cpp RayResultCallback .
- contactTest behaves like castRay. It uses group btBroadphaseProxy::DefaultFilter and is set to collide with everything (btBroadphaseProxy::AllFilter). Default Bullet behavior. See /usr/include/bullet/BulletCollision/CollisionDispatch/btCollisionWorld.h ContactResultCallback.

## How Has This Been Tested

Ad hoc testing with Andrew Z's pvizplan object-rearrangement project.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/esp/physics/bullet/BulletBase.h [hunk 1]
```diff
@@ -135,10 +135,9 @@ class BulletBase {
   std::shared_ptr<btMultiBodyDynamicsWorld> bWorld_;
 
   /** @brief Static data: All components of a @ref RigidObjectType::SCENE are
-   * stored here. See @ref btCollisionObject.  Also, all objects set to STATIC
-   * are stored here.
+   * stored here. Also, all objects set to STATIC are stored here.
    */
-  std::vector<std::unique_ptr<btCollisionObject>> bStaticCollisionObjects_;
+  std::vector<std::unique_ptr<btRigidBody>> bStaticCollisionObjects_;
```
Comment (eundersander) [inline src/esp/physics/bullet/BulletBase.h:140]: Duplicating a PR that went into master: https://github.com/facebookresearch/habitat-sim/pull/814

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 2]
```diff
@@ -144,7 +144,9 @@ bool BulletRigidObject::initialization_LibSpecific(
   bObjectRigidBody_ = std::make_unique<btRigidBody>(info);
 
   //! Add to world
-  bWorld_->addRigidBody(bObjectRigidBody_.get());
+  bWorld_->addRigidBody(bObjectRigidBody_.get(),
```
Comment (eundersander) [inline src/esp/physics/bullet/BulletRigidObject.cpp:147]: I'll explain my proposed collision-filtering in the main PR description.

File: src/esp/physics/bullet/BulletBase.h (hunk 1 above)
Comment (Skylion007) [reply src/esp/physics/bullet/BulletBase.h:140]: @eundersander Pull from master to make it disappear on the files changed tab.

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 3]
```diff
@@ -281,16 +285,18 @@ bool BulletRigidObject::setMotionType(MotionType mt) {
...
-    std::unique_ptr<btCollisionObject> staticCollisionObject =
-        std::make_unique<btCollisionObject>();
-    staticCollisionObject->setCollisionShape(bObjectShape_.get());
-    staticCollisionObject->setWorldTransform(
-        bObjectRigidBody_->getWorldTransform());
-    bWorld_->addCollisionObject(
-        staticCollisionObject.get(),
-        2,       // collisionFilterGroup (2 == StaticFilter)
-        1 + 2);  // collisionFilterMask (1 == DefaultFilter, 2==StaticFilter)
+    // mass == 0 to indicate static. See isStaticObject assert below. See also
```
Comment (eundersander) [inline src/esp/physics/bullet/BulletRigidObject.cpp:288]: This new code is mostly copied from https://github.com/facebookresearch/habitat-sim/pull/814, except notice also the new collision filtering.

File: src/esp/physics/bullet/BulletURDFImporter.cpp [hunk 4]
```diff
@@ -570,33 +570,6 @@ Mn::Matrix4 BulletURDFImporter::ConvertURDF2BulletInternal(
       bool isDynamic =
           (mbLinkIndex < 0 && cache.m_bulletMultiBody->hasFixedBase()) ? false
                                                                        : true;
-      int collisionFilterGroup = isDynamic
```
Comment (eundersander) [inline src/esp/physics/bullet/BulletURDFImporter.cpp:573]: This block is being moved below. Beware I'm also changing the collision filtering in the new version below.

File: src/esp/physics/bullet/BulletURDFImporter.cpp [hunk 5]
```diff
@@ -642,11 +616,38 @@ Mn::Matrix4 BulletURDFImporter::ConvertURDF2BulletInternal(
...
             col->setCollisionFlags(col->getCollisionFlags() |
                                    btCollisionObject::CF_STATIC_OBJECT);
+            isDynamic = false;
           }
         }
 
         cache.m_bulletMultiBody->setBaseCollider(col);
       }
+
+      int collisionFilterGroup = isDynamic
```
Comment (eundersander) [inline src/esp/physics/bullet/BulletURDFImporter.cpp:626]: Notice new collision filtering here.

File: src/esp/physics/bullet/BulletBase.h (hunk 1 above)
Comment (eundersander) [reply src/esp/physics/bullet/BulletBase.h:140]: Are you suggesting rebasing onto master? That's a bigger project than I'd like to tackle here.

**Comment by eundersander** [comment]
Note about CI failures: I don't think this branch is currently passing CI as-is (without my change).

**Review by aclegg3 — APPROVED** [review]
Looks good. I'll do a rebase to pull in other recent changes soon.

**Pull request merged**